control/canonical.py

```diff
@@ -149,97 +153,294 @@ def observable_form(xsys):
 
     return zsys, Tzx
 
-def modal_form(xsys):
-    """Convert a system into modal canonical form
+
+def similarity_transform(xsys, T, timescale=1, inverse=False):
+    """Perform a similarity transformation, with option time rescaling.
+
+    Transform a linear state space system to a new state space representation
+    z = T x, or x = T z, where T is an invertible matrix.
 
     Parameters
     ----------
     xsys : StateSpace object
-        System to be transformed, with state `x`
+           System to transform
+    T : 2D invertible array
+        The matrix `T` defines the new set of coordinates z = T x.
+    timescale : float
+        If present, also rescale the time unit to tau = timescale * t
+    inverse: boolean
+        If True (default), transform so z = T x.  If False, transform
+        so x = T z.
 
     Returns
     -------
     zsys : StateSpace object
-        System in modal canonical form, with state `z`
-    T : matrix
-        Coordinate transformation: z = T * x
-    """
-    # Check to make sure we have a SISO system
-    if not issiso(xsys):
-        raise ControlNotImplemented(
-            "Canonical forms for MIMO systems not yet supported")
+        System in transformed coordinates, with state 'z'
 
+    """
     # Create a new system, starting with a copy of the old one
     zsys = StateSpace(xsys)
 
-    # Calculate eigenvalues and matrix of eigenvectors Tzx,
-    eigval, eigvec = eig(xsys.A)
+    # Define a function to compute the right inverse (solve x M = y)
+    def rsolve(M, y):
+        return transpose(solve(transpose(M), transpose(y)))
 
-    # Eigenvalues and corresponding eigenvectors are not sorted,
-    # thus modal transformation is ambiguous
-    # Sort eigenvalues and vectors from largest to smallest eigenvalue
-    idx = eigval.argsort()[::-1]
-    eigval = eigval[idx]
-    eigvec = eigvec[:,idx]
+    # Update the system matrices
+    if not inverse:
+        zsys.A = rsolve(T, dot(T, zsys.A)) / timescale
+        zsys.B = dot(T, zsys.B) / timescale
+        zsys.C = rsolve(T, zsys.C)
+    else:
+        zsys.A = solve(T, zsys.A).dot(T) / timescale
+        zsys.B = solve(T, zsys.B) / timescale
+        zsys.C = zsys.C.dot(T)
+
+    return zsys
+
+
+_IM_ZERO_TOL = np.finfo(np.float64).eps ** 0.5
+_PMAX_SEARCH_TOL = 1.001
+
+
+def _bdschur_defective(blksizes, eigvals):
+    """Check  for defective modal decomposition
+
+    Parameters
+    ----------
+    blksizes: size of Schur blocks
+    eigvals: eigenvalues
```

Comment:
It's a private function, but use numpydoc anyway ( `blksizes : array_like of ...`)?